File: 21.merge-two-sorted-lists.js
/*
 * @lc app=leetcode id=21 lang=javascript
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
  if (!l1 && !l2) return null
  if (!l1) return l2
  if (!l2) return l1

  let head = currNode = new ListNode()
  while (l1 || l2) {
    if (!l1 || (l2 && l1.val > l2.val)) {
      currNode.next = new ListNode(l2.val)
      l2 = l2.next
    } else {
      currNode.next = new ListNode(l1.val)
      l1 = l1.next
    }
    currNode = currNode.next
  }
  return head.next
};
// @lc code=end

File: cpp/21.merge-two-sorted-lists.cpp
/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        // iterate through both lists together
        // add smaller value to new linked list
        // O(max(n, m)) time O(n+m) space
        ListNode *head = new ListNode();
        ListNode *currNode = new ListNode();
        head->next = currNode;
        while (list1 || list2)
        {
            if (!list1 || (list2 && list1->val > list2->val))
            {
                ListNode *newNode = new ListNode(list2->val);
                currNode->next = newNode;
                currNode = currNode->next;
                list2 = list2->next;
            }
            else if (!list2 || (list1 && list2->val >= list1->val))
            {
                ListNode *newNode = new ListNode(list1->val);
                currNode->next = newNode;
                currNode = currNode->next;
                list1 = list1->next;
            }
        }
        return head->next->next;
    }
};
// @lc code=end
